tests/test_integration_layout.py

```diff
@@ -0,0 +1,171 @@
+            # #root's virtual height should be as high as its stacked content
+            (SCREEN_W - 2 - 1, PLACEHOLDERS_DEFAULT_H * 4),
+            # placeholders width=same than screen, minus 2 borders, minus scrollbar :: height=default height minus 2 borders
+            (SCREEN_W - 2 - 1, PLACEHOLDERS_DEFAULT_H),
+            # placeholders should be at offset 1 because of #root's border
+            1,
+        ],
+        [
+            SCREEN_SIZE,
+            1,
+            "border: solid white;",  # #root has a visible border
+            "align: center top;",  # placeholders are centered horizontally
+            # #root's virtual size=screen size
+            (SCREEN_W, SCREEN_H),
+            # placeholders width=same than screen, minus 2 borders :: height=default height
+            (SCREEN_W - 2, PLACEHOLDERS_DEFAULT_H),
+            # placeholders should be at offset 1 because of #root's border
+            1,
+        ],
+        [
+            SCREEN_SIZE,
+            4,
+            "border: solid white;",  # #root has a visible border
+            "align: center top;",  # placeholders are centered horizontally
+            # #root's virtual height should be as high as its stacked content
+            (SCREEN_W - 2 - 1, PLACEHOLDERS_DEFAULT_H * 4),
+            # placeholders width=same than screen, minus 2 borders, minus scrollbar :: height=default height
+            (SCREEN_W - 2 - 1, PLACEHOLDERS_DEFAULT_H),
+            # placeholders should be at offset 1 because of #root's border
+            1,
+        ],
+    ),
+)
+async def test_vertical_container_with_children(
+    screen_size: Size,
+    placeholders_count: int,
+    root_container_style: str,
+    placeholders_style: str,
+    expected_placeholders_size: tuple[int, int],
+    expected_root_widget_virtual_size: tuple[int, int],
+    expected_placeholders_offset_x: int,
+    event_loop: asyncio.AbstractEventLoop,
+):
+    class VerticalContainer(Widget):
+        CSS = (
+            """
+        VerticalContainer {
+            layout: vertical;
+            overflow: hidden auto;
+            ${root_container_style}
+        }
+        
+        VerticalContainer Placeholder {
+            height: ${placeholders_height};
+            ${placeholders_style}
+        }
+        """.replace(
+                "${root_container_style}", root_container_style
+            )
+            .replace("${placeholders_height}", str(PLACEHOLDERS_DEFAULT_H))
+            .replace("${placeholders_style}", placeholders_style)
+        )
+
+    class MyTestApp(AppTest):
+        def compose(self) -> ComposeResult:
+            placeholders = [
+                Placeholder(id=f"placeholder_{i}", name=f"Placeholder #{i}")
+                for i in range(placeholders_count)
+            ]
+
+            yield VerticalContainer(*placeholders, id="root")
+
+    app = MyTestApp(size=screen_size, test_name="compositor")
+
+    async with app.in_running_state():
```

Comment:
this context manager starts the App, let it run for a little while so things are initialised (it takes a few event cycles to have the initial layout properly painted for example), yields it, and then shuts it down